src/metabase/db/liquibase.clj

```diff
@@ -216,66 +219,92 @@
+
+(def ^:private ^:dynamic *lock-depth* 0)
 
 (defn run-in-scope-locked
   "Run function `f` in a scope on the Liquibase instance `liquibase`.
-  Liquibase scopes are used to hold configuration and parameters (akin to binding dynamic variables in
-  Clojure). This function initializes the database and the resource accessor which are often required."
+   Liquibase scopes are used to hold configuration and parameters (akin to binding dynamic variables in
+   Clojure). This function initializes the database and the resource accessor which are often required.
+   The underlying locks are re-entrant, so it is safe to nest these blocks."
   [^Liquibase liquibase f]
-  (let [database (.getDatabase liquibase)
-        ^LockService lock-service (.getLockService (LockServiceFactory/getInstance) database)
-        scope-objects {(.name Scope$Attr/database) database
+
+  (when (.hasChangeLogLock (lock-service liquibase))
+    ;; In production it's better to just take the lock again, it is re-entrant.
+    ;; Fail in dev and CI in order to clean up the code flow.
+    (when-not config/is-prod?
+      (throw (LockException. "Attempted to take a Liquibase lock, but we already are holding it."))))
+
+  (let [database      (.getDatabase liquibase)
+        scope-objects {(.name Scope$Attr/database)         database
                        (.name Scope$Attr/resourceAccessor) (.getResourceAccessor liquibase)}]
     (Scope/child ^Map scope-objects
                  (reify Scope$ScopedRunner
                    (run [_]
-                     (.waitForLock lock-service)
+                     (wait-for-migration-lock liquibase)
+                     (assert-locked liquibase)
                      (try
-                       (f)
+                       (binding [*lock-depth* (inc *lock-depth*)]
+                         (f))
                        (finally
-                         (.releaseLock lock-service))))))))
+                         (when (zero? *lock-depth*)
+                           (.releaseLock (lock-service liquibase))))))))))
+
+(defmacro with-scope-locked
+  "Run `body` in a scope on the Liquibase instance `liquibase`.
+   Liquibase scopes are used to hold configuration and parameters (akin to binding dynamic variables in
+   Clojure). This function initializes the database and the resource accessor which are often required.
+   The underlying locks are re-entrant, so it is safe to nest these blocks."
+  {:style/indent 1}
+  [liquibase & body]
+  `(run-in-scope-locked ~liquibase (fn [] ~@body)))
+
+(defn migrate-up-if-needed!
+  "Run any unrun `liquibase` migrations, if needed."
+  [^Liquibase liquibase ^DataSource data-source]
+  (log/info (trs "Checking if Database has unrun migrations..."))
+  (if (seq (unrun-migrations data-source))
+    (do
+     (log/info (trs "Database has unrun migrations. Checking if migration lock is taken..."))
+     (with-scope-locked liquibase
```

Comment:
not the scope of this PR but I think we can acquire the lock at the top and remove all these double checking for unrun migrations. wdyt?

Reply:
I agree. From the PR description:

> There is scope to make [the locking scope] even more coarse, but I think this is good for now.